catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -0,0 +1,512 @@
+            throws SourceUnavailableException, IngestException {
+        final List<Metacard> versionedMetacards = metacards.stream()
+                .filter(MetacardVersion::isNotVersion)
+                .map(metacard -> new MetacardVersion(metacard, action, getSubject.get()))
+                .collect(Collectors.toList());
+
+        if (versionedMetacards.isEmpty()) {
+            return null;
+        }
+
+        return getSystemSubject().execute(() -> catalogProvider().create(new CreateRequestImpl(
+                versionedMetacards)));
+    }
+
+    private List<ContentItem> getVersionedContent(Metacard root, List<ContentItem> contentItems,
+            MetacardVersion.Action versionAction, Map<String, Path> tmpContentPaths) {
+        String id = root.getId();
+        MetacardVersion rootVersion = new MetacardVersion(root, versionAction, getSubject.get());
+        Supplier<Stream<ContentItem>> relatedContent = () -> contentItems.stream()
+                .filter(ci -> ci.getId()
+                        .equals(id));
+
+        ContentItem rootItem = relatedContent.get()
+                .filter(ci -> !isNullOrEmpty(ci.getUri()))
+                .filter(ci -> !ci.getUri()
+                        .contains("#"))
+                .findAny()
+                .orElseThrow(() -> new RuntimeException(
+                        "Could not find root content item for: " + id));
+
+        List<ContentItem> derivedContent = relatedContent.get()
+                .filter(ci -> !ci.equals(rootItem))
+                .collect(Collectors.toList());
+
+        List<ContentItem> resultItems = new ArrayList<>();
+
+        // version root content item
+        long size = 0;
+        try {
+            size = rootItem.getSize();
+        } catch (IOException e) {
+            LOGGER.warn("Could not get size of item [{}].", rootItem.getId(), e);
+        }
+        resultItems.add(new ContentItemImpl(rootVersion.getId(),
+                Files.asByteSource(tmpContentPaths.get(rootItem.getId())
+                        .toFile()),
+                rootItem.getMimeTypeRawData(),
+                rootItem.getFilename(),
+                size,
+                rootVersion));
+
+        // version derived content items
+        for (ContentItem contentItem : derivedContent) {
+            size = 0;
+            try {
+                size = contentItem.getSize();
+            } catch (IOException e) {
+                LOGGER.warn("Could not get size of item [{}].", rootItem.getId(), e);
+            }
+            resultItems.add(new ContentItemImpl(rootVersion.getId(),
+                    contentItem.getQualifier(),
+                    new WrappedByteSource(contentItem),
+                    contentItem.getMimeTypeRawData(),
+                    contentItem.getFilename(),
+                    size,
+                    rootVersion));
+        }
+
+        return resultItems;
+    }
+
+    private Subject getSystemSubject() {
```

Comment:
It looks like you always call execute on getSystemSubject.  Maybe you should change this method to take a runnable to execute as the system historian.  Please also add JavaDoc explaining this method should only be used to do history related operations since this elevates the execution permissions.  For example, it could be used to modify history which is something that should not be allowed and would need to be audited if it was supported in the future.